ui/StatusQ/src/fastexpressionrole.cpp

```diff
@@ -0,0 +1,163 @@
+    \inherits SingleRole
+    \inqmlmodule StatusQ
+    \brief A custom role similar to (and based on) SFPM's ExpressionRole but
+    optimized to access only explicitely indicated roles.
+
+    A FastExpressionRole, similarly as \l ExpressionRole, is a \l ProxyRole
+    allowing to implement a custom role based on a javascript expression.
+    However in FastExpressionRole's expression context there are available only
+    roles explicitely listed in \l expectedRoles property:
+
+    \code
+    SortFilterProxyModel {
+       sourceModel: numberModel
+       proxyRoles: FastExpressionRole {
+           name: "c"
+           expression: model.a + model.b
+           expectedRoles: ["a", "b"]
+      }
+    }
+    \endcode
+
+    By accessing only needed roles, the performance is significantly better in
+    comparison to ExpressionRole, especially when the model has multiple
+    FastExpressionRole's.
+*/
+
+/*!
+    \qmlproperty expression FastExpressionRole::expression
+
+    See ExpressionRole::expression for details. Unline the original
+    ExpressionRole only roles explicely declared via expectedRoles are accessible.
+*/
+const QQmlScriptString& FastExpressionRole::expression() const
+{
+    return m_scriptString;
+}
+
+void FastExpressionRole::setExpression(const QQmlScriptString& scriptString)
+{
+    if (m_scriptString == scriptString)
+        return;
+
+    m_scriptString = scriptString;
+    updateExpression();
+
+    emit expressionChanged();
+    invalidate();
+}
+
+void FastExpressionRole::proxyModelCompleted(const QQmlSortFilterProxyModel& proxyModel)
+{
+    updateContext(proxyModel);
+}
+
+void FastExpressionRole::setExpectedRoles(const QStringList& expectedRoles)
+{
+    if (m_expectedRoles == expectedRoles)
+        return;
+
+    m_expectedRoles = expectedRoles;
+    emit expectedRolesChanged();
+
+    invalidate();
+}
+
+/*!
+    \qmlproperty list<string> FastExpressionRole::expectedRoles
+
+    List of role names intended to be available in the expression's context.
+*/
+const QStringList& FastExpressionRole::expectedRoles() const
+{
+    return m_expectedRoles;
+}
+
+QVariant FastExpressionRole::data(const QModelIndex& sourceIndex,
+                                  const QQmlSortFilterProxyModel& proxyModel)
+{
+    if (!m_scriptString.isEmpty()) {
+        QVariantMap modelMap;
+        QHash<int, QByteArray> roles = proxyModel.roleNames();
+
+        QQmlContext context(qmlContext(this));
+        auto addToContext = [&] (const QString &name, const QVariant& value) {
+            context.setContextProperty(name, value);
+            modelMap.insert(name, value);
+        };
+
+        for (auto it = roles.cbegin(); it != roles.cend(); ++it) {
+            auto name = it.value();
+
+            if (!m_expectedRoles.contains(name))
+                continue;
+
+            addToContext(name, proxyModel.sourceData(sourceIndex, it.key()));
+        }
+
+        addToContext(QStringLiteral("index"), sourceIndex.row() + 6);
```

Comment:
What's this magic number `6`? :)

Reply:
:man_facepalming: Test added and fixed. I have no idea how it appeared here :sweat_smile: